### PR TITLE
Components: promote g2 `Popover` as `Flyout`

### PR DESCRIPTION
**Pull request by ciampo**
## Description

- Promote the `Popover` component from the `components/src/ui/` folder into the `components/src` folder
- Rename the component to `Flyout`
- Re-arrange the internal file structure into separate folders for the `Flyout` and `FlyoutContent` components.
- Refactor the implementation to use Emotion's `styled` approach
- Add documentation to the README
- Add the component to `tsconfig` and the docs manifest

**Next steps in follow-up PRs**:

- Spend time investigating / testing the newly promoted `Flyout` component to make sure it is stable and covers all use cases of the current (legacy) `Popover`. A few ways of conducting this investigation are:
     - Use storybook stories
     - Replace all Gutenberg usages of the legacy `Popover` with the new `Flyout`
- Once stability is reached, add a deprecation notice to the legacy `Popover`

## How has this been tested?

- [ ] Make sure that the `@wordpress/components` package builds correctly
- [ ] Check the [`Flyout` component in locally-run Storybook](http://localhost:50240/?path=/story/components-experimental-flyout--default) and make sure it behaves the same way as the [g2 `Popover` component on the production branch](https://wordpress.github.io/gutenberg/?path=/story/g2-components-experimental-popover--default)

## Screenshots 

Screenshot 2021-07-02 at 14 00 21

## Types of changes

Refactor (moving and partially rewriting files)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- N/A I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/accessible-popover/utils.js
Comment (sarayourfriend) [inline packages/components/src/accessible-popover/utils.js]: Currently not possible until `compose` is typed 😞 It's copied here for now for that reason.

**Comment by sarayourfriend** [comment]
> Convert component in TypeScript

This isn't really necessary. If you end up with extra time it's fine to do but according to the strategy laid out in [this make post about TypeScript in Gutenberg](https://make.wordpress.org/core/2021/04/09/follow-up-to-the-native-typescript-proposal/) we shouldn't be re-writing to TypeScript unless it's necessary for expressing the types (which in this case it isn't).

We haven't decided what to do about `Card` yet. If we can make it work with the `components/src/card` version (not the `ui` version) then we should prefer that. If not then part of this work will necessarily be uplifting the `Card` component 😞

**Comment by gziolo** [comment]
Can we keep the name `Popover` and use a prop that would do the switching?

@diegohaz used such approach for the `Toolbar` component:

https://github.com/WordPress/gutenberg/blob/152420f8d3b49163d46be9e42111dbb1893eab3f/packages/components/src/toolbar/index.js#L29-L37

**Comment by ciampo** [comment]
> Can we keep the name `Popover` and use a prop that would do the switching?
> 
> @diegohaz used such approach for the `Toolbar` component:

I'm quite new to the g2 components work, and so I don't really have much context to give a good answer to this question.

I based my choice on what was agreed with @sarayourfriend (and on previous conversations like [this comment](https://github.com/WordPress/gutenberg/issues/29689#issuecomment-800567214)), but I'm definitely also up for adopting the same approach used for the `Toolbar` component.

cc @griffbrad

**Comment by diegohaz** [comment]
> Can we keep the name `Popover` and use a prop that would do the switching?
> 
> @diegohaz used such approach for the `Toolbar` component:
> 
> https://github.com/WordPress/gutenberg/blob/152420f8d3b49163d46be9e42111dbb1893eab3f/packages/components/src/toolbar/index.js#L29-L37


I think this is more complicated than the Toolbar component. The difference between the legacy toolbar and the current accessible toolbar enabled by the `label` prop is basically 3 props (`controls`, `isCollapsed`, `title`). With Popover, it looks like we'll have a completely different API, which would be a mess to handle especially if you take TypeScript into account.

Also, another motivation to keep the `Toolbar` name was that we already have a standardized `toolbar` role in HTML. But "popover" is not standard. I see no problem in having a separate component as this is a different component with different behaviors anyway.

Maybe we can discuss more the name though, like `PopoverDialog`, `Flyout`, `ContextualDialog`, `Popup` etc. `popup` is a name that's used in a few places in the ARIA spec already (for example, [`aria-haspopup`](https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaHasPopup), which is set on the popover trigger element), but it's as generic as "dialog".

**Comment by gziolo** [comment]
> Maybe we can discuss more the name though, like `PopoverDialog`, `Flyout`, `ContextualDialog`, `Popup` etc. `popup` is a name that's used in a few places in the ARIA spec already (for example, [`aria-haspopup`](https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaHasPopup), which is set on the popover trigger element), but it's as generic as "dialog".

@diegohaz, thank you for feedback. We can have completely new name if that makes more sense here. The old `Popover` is also accessible so prefixing a new one with `Accessible` could confuse the package users.

**Comment by sarayourfriend** [comment]
Agreed that the name `AccessiblePopover` is bad. For one, it implies the previous implementation was inaccessible, which is not the case. `Popup` seems to be a good option based on Diego's feedback but might not offer a clear relationship between the old `Popover` component and the new `Popup` component. Do y'all think that might be important or is a deprecation warning in `Popover` enough?

File: packages/components/src/accessible-popover/utils.js
Comment (ciampo) [reply packages/components/src/accessible-popover/utils.js]: Composed is now typed! Rebasing on top of trunk will fix this :)

**Comment by ciampo** [comment]
I just rebased this PR to include a few changes that got recently merged on `trunk`, in particular:

- The component dependencies (`Surface`, `Scrollable` and `Card`) have all been promoted / updated to g2 implementation
- The `compose` package is fully typed

---

@sarayourfriend , @gziolo and @diegohaz  — let's discuss the next steps for this PR:

- we all agreed that `AccessiblePopover` is not a good name for this component. Should we go with `Popup`?  Other alternatives presented by @diegohaz were `PopoverDialog`, `Flyout` and `ContextualDialog`
- what deprecation strategy should we adopt for the legacy `Popover` component? is a deprecation warning enough?
- should we look into replacing usages of the legacy `Popover` in Gutenberg with the g2 version, once it lands? (in separate PRs)

**Comment by diegohaz** [comment]
> * we all agreed that `AccessiblePopover` is not a good name for this component. Should we go with `Popup`?  Other alternatives presented by @diegohaz were `PopoverDialog`, `Flyout` and `ContextualDialog`

`Popup` is too generic I guess. I would go for `Flyout`.

> * what deprecation strategy should we adopt for the legacy `Popover` component? is a deprecation warning enough?
> * should we look into replacing usages of the legacy `Popover` in Gutenberg with the g2 version, once it lands? (in separate PRs)

I think a deprecation warning is enough, but I would only do that after we make sure this component is stable and covers all the current `Popover`'s use cases. We should be able to replace all the internal usages of `Popover` without a huge effort. If, for some reason, we can't migrate all `Popover` in Gutenberg or if it's too challenging, I don't think the `Popover` component should be deprecated.

**Comment by ciampo** [comment]
> `Popup` is too generic I guess. I would go for `Flyout`.

Alright, `Flyout` seems like a better option to me as well.

> I think a deprecation warning is enough, but I would only do that after we make sure this component is stable and covers all the current Popover's use cases. We should be able to replace all the internal usages of Popover without a huge effort. If, for some reason, we can't migrate all Popover in Gutenberg or if it's too challenging, I don't think the Popover component should be deprecated.

This also makes sense. It looks like the next steps could be:

1. Promote g2 `Popover` as `Flyout` (and keep it marked as _experimental_)
2. Spend time investigating / testing the newly promoted `Flyout` component to make sure it is stable and covers all use cases of the current (legacy) `Popover`, and apply any amends if necessary in follow-up PRs. A few ways of conducting this investigation are:
     - Use storybook stories
     - Replace all Gutenberg usages of the legacy `Popover` with the new `Flyout`
3. Once stability is reached, add a deprecation notice to the legacy `Popover`

File: packages/components/src/flyout/flyout/component.js
Comment (ciampo) [inline packages/components/src/flyout/flyout/component.js]: Not sure if any of this could be also moved to the hook

File: packages/components/src/flyout/index.js
Comment (ciampo) [inline packages/components/src/flyout/index.js]: Should the `FlyoutContent` component be exported in the public APIs of `@wordpress/components`?

File: packages/components/src/flyout/types.ts
Comment (ciampo) [inline packages/components/src/flyout/types.ts]: Should we also rename `popover` to `flyout` (or `flyoutState`)?

File: packages/components/src/flyout/flyout-content/component.js
Comment (ciampo) [inline packages/components/src/flyout/flyout-content/component.js]: Should this also be moved to the hook?

File: packages/components/src/flyout/flyout/README.md
Comment (sarayourfriend) [inline packages/components/src/flyout/flyout/README.md]:
```suggestion
import { Button, __experimentalFlyout as Flyout, __experimentalText as Text } from '@wordpress/components';
```

File: packages/components/src/flyout/index.js
Comment (sarayourfriend) [reply packages/components/src/flyout/index.js]: No. I'd argue that because it's just an internal component it shouldn't even be exported here either.

**Review by sarayourfriend — COMMENTED** [review]
LGTM so far!

File: packages/components/src/flyout/flyout-content/hook.js
Comment (sarayourfriend) [inline packages/components/src/flyout/flyout-content/hook.js]: We'll need to refactor FlyoutContent to use `@emotion/styled` and `@emotion/react`.

File: packages/components/src/flyout/flyout/component.js
Comment (ciampo) [reply packages/components/src/flyout/flyout/component.js]: Not an issue anymore, since we're going to used `styled`  — which moves us away from hooks

File: packages/components/src/flyout/flyout-content/component.js
Comment (ciampo) [reply packages/components/src/flyout/flyout-content/component.js]: Same as https://github.com/WordPress/gutenberg/pull/32197#discussion_r662060932

File: packages/components/src/flyout/types.ts
Comment (diegohaz) [reply packages/components/src/flyout/types.ts]: Yeah! I think that makes sense.

File: packages/components/src/flyout/flyout-content/hook.js
Comment (ciampo) [reply packages/components/src/flyout/flyout-content/hook.js]: Done in https://github.com/WordPress/gutenberg/pull/32197/commits/3624b7794b3b095d05619ee4053e86b81abc1d53 and https://github.com/WordPress/gutenberg/pull/32197/commits/3598b735319eaedc5b498e7d470a3b9f2d527243

File: packages/components/src/flyout/flyout/README.md
Comment (ciampo) [reply packages/components/src/flyout/flyout/README.md]: Done in https://github.com/WordPress/gutenberg/pull/32197/commits/77b5734b35b36164c217029c347363391cf69331

File: packages/components/src/flyout/types.ts
Comment (ciampo) [reply packages/components/src/flyout/types.ts]: Done in https://github.com/WordPress/gutenberg/pull/32197/commits/8bd3af8656f462cafd26871e439d38ca53bd6eae

File: packages/components/src/flyout/flyout-content/component.js
Comment (ciampo) [inline packages/components/src/flyout/flyout-content/component.js:10]: The refactoring done in https://github.com/WordPress/gutenberg/pull/32197/commits/3598b735319eaedc5b498e7d470a3b9f2d527243, (where we wrapped `ReakitPopover` directly into the `styled` function), caused a typescript error around the `as` prop on `< FlyoutContentView />`.

For this reason, `FlyoutContent` (and `Flyout`) have been marked as non-polymorphic

**Comment by ciampo** [comment]
This PR is now ready for a full review! I've also updated the description to match the latest changes. @sarayourfriend and @diegohaz , would you mind taking another look? Thank you!

File: packages/components/src/flyout/flyout/component.js
Comment (sarayourfriend) [inline packages/components/src/flyout/flyout/component.js]: Just to clarify, why is `Flyout` itself non-polymorphic?

Comment (ciampo) [reply packages/components/src/flyout/flyout/component.js]: It's because TypeScript would otherwise complain when expanding `{ ...otherProps }` into `<FlyoutComponent />`.

Since the `as` prop from `Flyout` can't be applied to `FlyoutContent` (since `FlyoutContent` is non-polymorphic), I thought that `Flyout` itself should also be marked as non-polymorphic.

Do you have an alternative approach to this?

Comment (sarayourfriend) [reply packages/components/src/flyout/flyout/component.js]: Oh I see, the as prop would be applied to FlyoutContent, not some wrapper in Flyout itself? This makes sense then.

File: packages/components/src/flyout/flyout-content/component.js
Comment (diegohaz) [reply packages/components/src/flyout/flyout-content/component.js:10]: Out of curiosity, what's the context behind refactoring the `css` calls to `styled.div` and then to `styled( ReakitPopover )`? I remember we've had some conversations about that, but I can't remember exactly the reasoning behind it. cc @sarayourfriend

Comment (sarayourfriend) [reply packages/components/src/flyout/flyout-content/component.js:10]: Basically `@emotion/css` does not hook into the Emotion cache so it's not possible for us to use it's `css` function because we need to hook into the cache to support iframes.

`@emotion/styled` does hook into the CacheProvider so it is able to support iframes.

Comment (diegohaz) [reply packages/components/src/flyout/flyout-content/component.js:10]: Ah yeah! That makes sense.

File: packages/components/src/flyout/index.js
Comment (sarayourfriend) [inline packages/components/src/flyout/index.js]: Do we need to export the `context`?

File: docs/manifest.json
Comment (sarayourfriend) [inline docs/manifest.json]: I think this should be removed if we're not exporting it right?

Comment (ciampo) [reply docs/manifest.json]: I believe that in order to remove these lines, we should also delete `/packages/components/src/flyout/flyout-content/README.md` — should I go ahead and delete that file too?

Comment (sarayourfriend) [reply docs/manifest.json]: Wow, that's pretty wild.

My concern is that we're adding external documentation for an internal component. I guess we should remove the README? 🤷

File: packages/components/src/flyout/index.js
Comment (ciampo) [reply packages/components/src/flyout/index.js]: I kept that export since it was also in the "original" g2 popover component, but it's actually not used by any external package. I removed it in https://github.com/WordPress/gutenberg/pull/32197/commits/acdc90605ff14f88358bd8e8fac3626a04306fa7

File: docs/manifest.json
Comment (ciampo) [reply docs/manifest.json]: I tried removing only the lines in `docs/manifest.json` in https://github.com/WordPress/gutenberg/pull/32197/commits/aa02f1e13033bd5627a1bad00b11cec84ed14d9e — let's see if the CI tasks still pass, otherwise I will also delete the README

Comment (ciampo) [reply docs/manifest.json]: Yep, the Static Analysis task is failing.

Screenshot 2021-07-02 at 19 58 38


I went ahead and removed the README file in https://github.com/WordPress/gutenberg/pull/32197/commits/35dfc6bca8f1a6105b371a57f20f135a522e0ee9

**Review by youknowriad — COMMENTED** [review]
Hi there!, thanks for the PR, I'd like to understand the plan for the future here? Is the idea to have a complete separate component, replace the existing one in the end, how do you see this?

Also, this has an impact on all the components that rely on it (like Dropdown, DropdownMenu...) and that supports popoverProps, will we need a separate component for all of these? How are we approaching this? 

Thanks for the clarifications.

**Comment by ciampo** [comment]
Hey @youknowriad ,

[this](https://github.com/WordPress/gutenberg/pull/32197#issuecomment-868486079) is the agreed plan of action between @sarayourfriend , @diegohaz , @gziolo  and myself:

> 1. Promote g2 `Popover` as `Flyout` (and keep it marked as _experimental_)
> 2. Spend time investigating / testing the newly promoted `Flyout` component to make sure it is stable and covers all use cases of the current (legacy) `Popover`, and apply any amends if necessary in follow-up PRs. A few ways of conducting this investigation are:
>     - Use storybook stories
>     - Replace all Gutenberg usages of the legacy `Popover` with the new `Flyout`
> 3. Once stability is reached, add a deprecation notice to the legacy `Popover`

This PR represents step 1, and is creating a new, separate component.

We will then follow up and start adding Storybook stories and replacing usages of the legacy popover in Gutenberg with the new `Flyout` component — which will also help us finding and ironing our any issues with the current `Flyout` component (which is going to be marked as experimental).

With regards to your question:

> Also, this has an impact on all the components that rely on it (like Dropdown, DropdownMenu...) and that supports popoverProps, will we need a separate component for all of these? How are we approaching this?

I assume that this will be discussed while working on phase 2. The main reason for creating a new, separate component (instead of replacing the current one) is that the APIs are very different.

One solution could be to write some logic that understands when the old APIs are being used, and _translates_ the old Popover Props to the new props automatically. We could update the components' documentation and remove mentions of the old popover props. Finally, we could add a deprecation warning when the old props are used.

Once we're done with step 2, we will deprecate the legacy `Popover` component altogether (and discuss whether it should be kept around for compact reasons or removed altogether).

**Comment by youknowriad** [comment]
> I assume that this will be discussed while working on phase 2. The main reason for creating a new, separate component (instead of replacing the current one) is that the APIs are very different.

It would be great to have at least an idea of where we want to go here. Personally, I think it's not great to have to create alternate components for all of these (Dropdown...), it seems too much which raises the question, why a separate component for just Popover, why not just to have a condition on the Popover component to switch implementation and translate old props to new ones (with deprecations), which would prevent us from changing the usage of Popover in all these other components. I think at least having a plan on how to handle these now would be good.

**Comment by ciampo** [comment]
> why a separate component for just Popover, why not just to have a condition on the Popover component to switch implementation and translate old props to new ones (with deprecations), which would prevent us from changing the usage of Popover in all these other components

@diegohaz's [previous answer](https://github.com/WordPress/gutenberg/pull/32197#issuecomment-858744406) highlights quite well the main reason why we agreed on having `Flyout` as a separate component.

>  Personally, I think it's not great to have to create alternate components for all of these (Dropdown...)

That would also be my preference, but I'm curious to hear what everyone else involved in this conversation thinks too.

**Comment by sarayourfriend** [comment]
> Personally, I think it's not great to have to create alternate components for all of these (Dropdown...),

That was never the plan 🙂 The part from the #30503 issue that Marco quotes states that we plan on replacing all the usages of Popover in `@wordpress/components` with Flyout. And yes, that will involve deprecating props as you suggested.

**Comment by youknowriad** [comment]
> we plan on replacing all the usages of Popover in @wordpress/components with Flyout.

I think what I missed there is that we can't just do that since the APIs of `Popover` are part of the APIs of these components as well. Meaning it's not just `Popover` that we're deprecating but we'd also need to maintain a condition in these components to fallback to `Popover` if these props are used. 

Which brings me to: why have this condition duplicated across all of these components instead of just keeping `Popover` and have the condition inside it.

**Comment by sarayourfriend** [comment]
> why have this condition duplicated across all of these components instead of just keeping Popover and have the condition inside it.

I think it's better to have the separate implementations for the reasons that Diego shared in the comment that Marco linked.

If we find that we're duplicate conditions across the `@wordpress/components` codebase, it would be trivial to abstract those either into an internal `PopoverOrFlyout` utility component that delegates to the correct one based on props, or some other function that does the same thing. We can always abstract the logic out.

This has at least one benefit of not further complicating the Popover component, which is already quite complex. We can contain that complexity outside of either component, allowing Flyout to be used directly without any intermediary step while still keeping the transition ergonomic for us within the components library.

**Comment by youknowriad** [comment]
I don't know I feel like having `popover/legacy` and `popover/new` inside the Popover component's folder would allow us to both separate implementations and keep a single entry point from the get go. It makes it clear that these are just the same component being migrated and potentially when we reach the final stage `popover/legacy` could be removed or could just include the `oldPropsToNewProps` function + deprecation (without the component itself).

**Comment by sarayourfriend** [comment]
Given the API differences I'm not sure there's a viable `oldPropsToNewProps` in this situation. I've thought about this many times and was unable to come up with any way of reconciling the clientRect concept to the trigger concept in Popover and Flyout respectively. If someone is able to solve this problem, however, that would be incredible and indeed would mean we could delete the old Popover and keep the name the same (which is obviously preferred, it's what we've done for Text and Flex and others).

**Comment by youknowriad** [comment]
To circle back a bit, I guess I should clarify something. What matters the most is the end result :) 

Is it better to tell all folks to switch from `Popover` from `Flyout` because some props might have changed or to only deprecate some props (and not the whole component) which means some users might not have to change anything will others might have to do so.

I think I'd prefer to just keep the `Popopver` component personally. For the current PR, it's still about an experimental API, which mean I wouldn't mind an experimental component but it feels like it's going to have impact on the final result :)

**Comment by youknowriad** [comment]
Ok, just saw the last comment and it's fair, I think we might not have a migration for all cases and we'd have to keep the fallback anyway but maybe it's only for a subset of use-cases?

**Comment by youknowriad** [comment]
Thinking more here, do we agree that for now, it seems impossible to make the best decision on whether this should be a separate component or not (in the end) because we didn't implement all the required features from Popover in Flyout yet (our own use-cases).

**Comment by sarayourfriend** [comment]
> Thinking more here, do we agree that for now, it seems impossible to make the best decision on whether this should be a separate component or not (in the end) because we didn't implement all the required features from Popover in Flyout yet (our own use-cases).

We can wait to export Flyout until the use cases are covered, if that's what you're suggesting? Sorry, just trying to make sure I understand the actionable part of your suggestion 🙂

**Comment by youknowriad** [comment]
yes exactly, I feel (maybe I'm alone though) that potential migration and deprecation will inform us better on whether this should be a separate component or not. In other words:

 - I'm fine with this PR since it doesn't impact anyone and just adds an experimental component
 - when we start using `Flyout` in other places, we should be mindful about API impact (Things like Dropdown...)
 - Replacing Popover usage means implementing the missing features into Flyout and be in a better position to decide whether they can be merged

**Comment by diegohaz** [comment]
> Thinking more here, do we agree that for now, it seems impossible to make the best decision on whether this should be a separate component or not (in the end) because we didn't implement all the required features from Popover in Flyout yet (our own use-cases).

Yeah! I agree with that. I don't think we'll have a final answer before we apply it to different use cases and find out how many things will break. My only certainty is that things will break. 😅 

Furthermore, in case we find out that it doesn't work at all or it is not worth the effort, I think it's easier to revert the experiment if it's a separate component.

**Comment by ciampo** [comment]
I've rebased the PR and all tests are now passing.

From what I gathered, we seem to agree that the next steps are:

- Merge this PR (which introduces `Flyout` as a new, experimental component)
- Look into replacing usages of `Popover` with `Flyout` in Gutenberg, and gather knowledge on any issues/incompatibilities that are caused by the potential swap
- Discuss the next steps based on the knowledge gathered 

If this is the case, we need a final review/approval before this PR can be merged.

**Review by sarayourfriend — APPROVED** [review]
LGTM! 🚀 Storybook tests well and the code makes sense. Thanks for your persistance on this very long one!

**Pull request merged**
